### PR TITLE
JuMP/MOI: Basic support for quadratic expressions and constraints

### PR DESCRIPTION
**Pull request by mlubin**
~Quadratic objectives don't yet work because of https://github.com/JuliaOpt/MathOptInterfaceUtilities.jl/issues/12~

- Renamed `LinearConstraint` to `AffExprConstraint`
- Similarly, `QuadExprConstraint`

### CODE REVIEW

**Comment by mlubin** [comment]
This is updated and ready to go.

File: src/operators.jl [hunk 1]
```diff
@@ -87,9 +87,17 @@ if VERSION < v"0.6.0-dev.1632" # julia PR #17623
...
 end
 # AffExpr--Variable
-(+){C,V<:JuMPTypes}(lhs::GenericAffExpr{C,V}, rhs::V) = (+)(rhs,lhs)
+(+){C,V<:JuMPTypes}(lhs::GenericAffExpr{C,V}, rhs::V) = GenericAffExpr{C,V}(vcat(lhs.vars,rhs),vcat(lhs.coeffs,one(C)), lhs.constant)
 (-){C,V<:JuMPTypes}(lhs::GenericAffExpr{C,V}, rhs::V) = GenericAffExpr{C,V}(vcat(lhs.vars,rhs),vcat(lhs.coeffs,-one(C)),lhs.constant)
-(*)(lhs::AffExpr, rhs::Variable) = (*)(rhs,lhs)
+# Don't fall back on AffExpr*Variable to preserve lhs/rhs consistency (appears in printing)
+function (*)(lhs::AffExpr, rhs::Variable)
+    n = length(lhs.vars)
+    if lhs.constant != 0.
```
Comment (blegat) [inline src/operators.jl]: Could we do simply `lhs.constant * rhs` here and move the `if` to `(*)(lhs::Number, rhs::Variable)` ?

File: src/quadexpr.jl [hunk 2]
```diff
@@ -61,25 +60,59 @@ function Base.isequal{T,S}(q::GenericQuadExpr{T,S},other::GenericQuadExpr{T,S})
...
+    return MOI.ScalarQuadraticFunction(instancereference.(q.aff.vars), q.aff.coeffs, instancereference.(q.qvars1), instancereference.(q.qvars2), q.qcoeffs, q.aff.constant)
+end
+
+function QuadExpr(m::Model, f::MOI.ScalarQuadraticFunction)
+    return QuadExpr(Variable.(m,f.quadratic_rowvariables), Variable.(m, f.quadratic_colvariables), f.quadratic_coefficients, AffExpr(Variable.(m,f.affine_variables), f.affine_coefficients, f.constant))
+end
+
+function setobjective(m::Model, sense::Symbol, a::QuadExpr)
+    @assert !m.solverinstanceattached # TODO
+    if sense == :Min
```
Comment (blegat) [inline src/quadexpr.jl]: Why don't we do
```julia
function setobjective(m::Model, sense::Symbol, a::AbstractJuMPScalar)
    ...
    MOI.set!(m.instance, MOI.ObjectiveFunction(), JuMPScalar2MOIfunction(a))
    nothing
end
```
Of course the name `JuMPScalar2MOIfunction` should be replaced by something better.

File: src/quadexpr.jl [hunk 3]
```diff
@@ -61,25 +60,59 @@ function Base.isequal{T,S}(q::GenericQuadExpr{T,S},other::GenericQuadExpr{T,S})
...
-#             MathProgBase.setquadobjterms!(m.internalModel, Cint[v.col for v in m.obj.qvars1], Cint[v.col for v in m.obj.qvars2], m.obj.qcoeffs)
-#         else
-#             isa(m.internalModel, MathProgBase.AbstractLinearQuadraticModel) && Base.warn_once("Solver does not support adding a quadratic objective to an existing model. JuMP's internal model will be discarded.")
-#             m.internalModelLoaded = false
-#         end
-#     end
-#     setobjectivesense(m, sense)
-# end
+function MOI.ScalarQuadraticFunction(q::QuadExpr)
+    return MOI.ScalarQuadraticFunction(instancereference.(q.aff.vars), q.aff.coeffs, instancereference.(q.qvars1), instancereference.(q.qvars2), q.qcoeffs, q.aff.constant)
```
Comment (joehuchette) [inline src/quadexpr.jl]: What is ``instancereference`` here?

Comment (joehuchette) [reply src/quadexpr.jl]: ohhhh is this a vectorized constructor? It looked like a crazy field access...

File: src/quadexpr.jl [hunk 4]
```diff
@@ -105,67 +138,27 @@ end
...
-const QuadConstraint = GenericQuadConstraint{QuadExpr}
+"""
+    addconstraint(m::Model, c::QuadExprConstraint)
 
-function Base.copy(c::QuadConstraint, new_model::Model)
-    return QuadConstraint(copy(c.terms, new_model), c.sense)
+Add a `QuadExpr` constraint to `Model m`.
+"""
+function addconstraint(m::Model, c::QuadExprConstraint)
+    @assert !m.solverinstanceattached # TODO
```
Comment (joehuchette) [inline src/quadexpr.jl]: This will be addressed when problem modification support is added?

File: src/quadexpr.jl (hunk 3 above)
Comment (mlubin) [reply src/quadexpr.jl]: Yes, welcome to the future

File: src/quadexpr.jl (hunk 4 above)
Comment (mlubin) [reply src/quadexpr.jl]: Definitely

File: src/operators.jl (hunk 1 above)
Comment (mlubin) [reply src/operators.jl]: How? What is `(*)(lhs::Number, rhs::Variable)` supposed to return if `lhs = 0`? An `AffExpr` with zero terms?

Comment (mlubin) [reply src/operators.jl]: Ok, `zero(AffExpr)` makes sense and is arguably simpler. However, I don't want to make arbitrary tweaks in performance critical code without profiling. This new method just copies Variable*AffExpr.

File: src/quadexpr.jl (hunk 2 above)
Comment (mlubin) [reply src/quadexpr.jl]: Makes sense, but I'll leave it for a future refactor.

**Pull request merged**